code/iaas/allocator/src/main/java/io/cattle/platform/allocator/constraint/DiskSizeConstraint.java

```diff
@@ -0,0 +1,96 @@
+
+    private static final Logger log = LoggerFactory.getLogger(DiskSizeConstraint.class);
+
+    private Long reserveSize;
+    private String volumeName;
+    private ObjectManager objectManager;
+
+    public DiskSizeConstraint(String volumeName, String sizeWithUnit, ObjectManager objMgr) {
+        this.volumeName = volumeName;
+        this.reserveSize = Long.parseLong(sizeWithUnit.replaceAll("[^0-9]", ""));
+        this.objectManager = objMgr;
+    }
+
+    @Override
+    public boolean matches(AllocationAttempt attempt, AllocationCandidate candidate) {
+        Set<Long> hostIds = candidate.getHosts();
+
+        // if one of the host does not have enough free space then return false
+        for (Long hostId : hostIds) {
+            boolean oneGood = false;
+
+            CacheManager cm = CacheManager.getCacheManagerInstance(this.objectManager);
+            HostInfo hostInfo = cm.getHostInfo(hostId);
+
+            // we will get a bunch of disks for that host and we need at least one disk with
+            // free space large enough
+            oneGood = checkDiskSize(cm, hostInfo, attempt.getInstanceId());
+
+            // if no disk with big enough free space for this host, then
+            // candidate is no good
+            if (!oneGood) {
+                log.debug("Scheduling instance [{}] to host [{}] rejected", attempt.getInstanceId(), hostId);
+                return false;
+            }
+        }
+        return true;
+    }
+
+    public boolean checkDiskSize (CacheManager cm, HostInfo hostInfo, Long instanceId) {
+
+        for (Entry<String, DiskInfo> entry : hostInfo.getAllDiskInfo()) {
+            DiskInfo diskInfo = entry.getValue();
+            Long capacity = entry.getValue().getCapacity();
+            Long allocatedSize = ((DiskInfo) diskInfo).getAllocatedSize();
+            Long freeSize = capacity - allocatedSize;
+
+            InstanceInfo instanceInfo = cm.getInstanceInfoForHost(hostInfo.getHostId(), instanceId);
+            
+            // we need to consider more than one volume constraints for this
+            // container case. That means free size will subtract all reserved
+            // size by other volumes as well beside allocated size.
+            InstanceDiskReserveInfo diskReserved = instanceInfo.getDiskReserveInfo(((DiskInfo)diskInfo).getDiskDevicePath());
+            if (diskReserved != null) {
+                freeSize -= diskReserved.getReservedSize();
+            }
+            if (freeSize >= this.reserveSize) {
+                // reserve disk for this instance
+                if (diskReserved != null) {
+                    diskReserved.addReservedSize(this.reserveSize);
+                } else {
+                    diskReserved = new InstanceDiskReserveInfo(((DiskInfo)diskInfo).getDiskDevicePath(), this.reserveSize, this.volumeName);
+                    instanceInfo.reserveDisk(diskReserved);
```

Comment:
I have a couple concerns about this logic:

##### Concern 1: You might have a memory leak
You are modifying the data in the cache here, right?
Let's say you have two candidate hosts: A and B. They both have enough free diskspae, so would they both get a cache entry reserving the disk space?

My understanding is that you use the reservation entry in the cache to then update the amount of available disk in allocatorDao.modifyDisk(). So, the fact that you have these bogus entries for every host in the environment doesn't affect the calculation, BUT you _only_ remove entries for where the host to which the instance was _actually_ scheduled (in InstancePurgePostListener), right?

So, you aren't cleaning up all entries in the cache for an instance when it is purged and you have a memory leak, right?


##### Concern 2: You are using the cache to temporarily persist state in memory so that it can be easily accessed later
The InstanceDiskReserveInfo entry's current purpose is to be accessible by allocatorDao.modifyDisk() so that that method can update its cached tally of disk space, right?

I don't think storing this data in memory is a good design. As evidence by concern #1, it can lead to things like memory leaks. We've tried to avoid this in cattle.

So, I think:
* AllocatorDao.modifyDisk() should just recalculate the disk needed when its called as opposed to pulling an entry from the cache manager. You could use a shared util method to ensure both places are calculating the same.
* The _only_ thing that should be stored in the cache is the running tally of available disk space for each host.